### PR TITLE
Implement accessors, reading methods and unit tests

### PR DESCRIPTION
**Pull request by ranchalp**

### CODE REVIEW

File: ipc_types/ipc_subnet_actor.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+func CheckpointEpoch(epoch, period ChainEpoch) ChainEpoch {
+	return (epoch / period) * period
+}
+
+func WindowEpoch(epoch, period ChainEpoch) ChainEpoch {
+	ind := epoch / period
+	return period * (ind + 1)
+}
+
+type CheckData struct {
```
Comment (adlrocha) [inline ipc_types/ipc_subnet_actor.go:170]: We are missing some methods for some of the checkpoint-related types (mainly the important is `Cid()`). Let's maybe add `Cid()` for the ones that implement it and we can add the pending ones in the future PRs once we kick-off the integration.

File: ipc_types/types.go [hunk 2]
```diff
@@ -4,185 +4,192 @@ package ipc_types
...
-
 type IPCAddress struct {
 	SubnetID   SubnetID
 	RawAddress address.Address
 }
 
-type SubnetID struct {
-	Parent string
-	Actor  address.Address
+func newIPCAddressFromID(id uint64) *IPCAddress {
```
Comment (adlrocha) [inline ipc_types/types.go:17]: We should remove this function, it leads to an invalid subnet address. This is not your fault, it is also the case in the Rust implementation, we will fix it there.

File: ipc_types/types.go [hunk 3]
```diff
@@ -4,185 +4,192 @@ package ipc_types
...
+	Actor:  id0,
+}
+
+// UndefSubnetID is the undef ID
+var UndefSubnetID = SubnetID{
+	Parent: UndefStr,
+	Actor:  id0,
+}
+
+func NewSubnetIDFromString(parent string, subnetAct address.Address) *SubnetID {
```
Comment (adlrocha) [inline ipc_types/types.go:75]: This one is wrong `NewSubnetIDFromString` should expect a `string` as an input. Again, not your fault, something to fix in Rust also, and it should implement the following logic: https://github.com/consensus-shipyard/ipc-actors/blob/2d2428eb37edc83155d8706e5c5f8ddb1f526ec8/sdk/src/subnet_id.rs#L187

File: ipc_types/types.go [hunk 4]
```diff
@@ -4,185 +4,192 @@ package ipc_types
...
+		Parent: parent,
+		Actor:  subnetAct,
+	}
+}
+
+func NewSubnetID(parent SubnetID, subnetAct address.Address) *SubnetID {
+	return NewSubnetIDFromString(parent.String(), subnetAct)
+}
+
+func (s SubnetID) ToBytes() []byte {
```
Comment (adlrocha) [inline ipc_types/types.go:86]: In go we use the convention of calling it `Bytes` (not `to_bytes` like in Rust).
```suggestion
func (s SubnetID) Bytes() []byte {
```

File: ipc_types/types.go [hunk 5]
```diff
@@ -4,185 +4,192 @@ package ipc_types
...
+	return []byte(strID)
+}
+
+// String returns the id in string form.
+func (id SubnetID) String() string {
+	if id == RootSubnet {
+		if id.Parent != UndefStr {
+			return RootStr
+		} else {
+			return UndefStr //TODO found bug in here perhaps? check with Alfonso
```
Comment (adlrocha) [inline ipc_types/types.go:97]: What bug?

**Review by adlrocha — CHANGES_REQUESTED** [review]
Did a first pass. I will add an issue for the things that are wrong in the Rust implementation and that need to change here and there (so you can track them). 

Also, it would be great if you can set-up CI here so tests are automatically checked in every PR. You can use Github Actions (like Denis did here: https://github.com/consensus-shipyard/spacenet/tree/main/.github/workflows) or CircleCI (as we do for the rest of PL repos). Ping Denis or me if you need help with this.

File: utils/utils.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+		_, err = adtMap.Get(k, i)
+	} else {
+		return nil, fmt.Errorf("the type *%T does not implement the cbor.Unmarshaler interface", out)
+	}
+	return &out, err
+}
+
+var (
+	threshold         mbig.Rat
+	calcThresholdOnce sync.Once
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply utils/utils.go:59]: The threshold is always 2/3 with the current implementation if I am not mistaken? but it is a minor optimization, changed it to calculating 2/3 (as an `mbig.rat`) every time instead.

File: COPYRIGHT [hunk 7]
```diff
@@ -0,0 +1,3 @@
+Copyright 2020. Protocol Labs, Inc.
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply COPYRIGHT:1]: Done

File: ipc_types/ipc_gateway.go [hunk 8]
```diff
@@ -0,0 +1,187 @@
+package ipc_types
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply ipc_types/ipc_gateway.go:1]: Done

File: utils/utils.go [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+	adtArray, err := adt.AsArray(s, cID, bitwidth)
+	var (
+		out   T
+		found bool
+	)
+
+	if i, ok := (any(&out)).(cbor.Unmarshaler); ok {
+		found, err = adtArray.Get(nonce, i)
+	} else {
+		return nil, fmt.Errorf("the type *%T does not implement the cbor.Unmarshaler interface", out)
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply utils/utils.go:28]: Done

File: utils/utils.go [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+		return nil, xerrors.Errorf("failed to get %T with nonce %v: %w", out, nonce, err)
+	}
+	if !found {
+		return nil, nil
+	}
+	return &out, nil
+}
+
+// GetOutOfHamt takes a generic type that must implement cbor.Unmarshaler
+// and returns a particular vale of a THamt type passed as cid.Cid given the key
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply utils/utils.go:40]: Done

File: ipc_types/ipc_gateway.go [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+	CircSupply     abi.TokenAmount
+	Status         Status
+	PrevCheckpoint Checkpoint
+}
+
+func (sn Subnet) GetTopDownMsg(s adt.Store, nonce uint64) (*CrossMsg, error) {
+	return utils.GetOutOfArray[CrossMsg](sn.TopDownMsgs, s, nonce, CrossMsgsAMTBitwidth)
+}
+
+func getTopDownMsg(crossMsgs *adt.Array, nonce uint64) (*CrossMsg, error) {
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply ipc_types/ipc_gateway.go:41]: Done

File: ipc_types/ipc_gateway.go [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+			return nil, err
+		}
+		if meta != nil { //then found
+			out = append(out, meta)
+		}
+	}
+	return out, nil
+}
+
+func getBottomUpMsgMeta(crossMsgs *adt.Array, nonce uint64) (*CrossMsgMeta, error) {
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply ipc_types/ipc_gateway.go:115]: done

File: ipc_types/ipc_subnet_actor.go [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+func (s *IPCSubnetActorState) IsValidator(addr *address.Address) bool {
+	for _, x := range s.ValidatorSet {
+		if x.Addr == *addr {
+			return true
+		}
+	}
+	return false
+}
+
+func getMsgMeta(msgMetas *adt.Map, c cid.Cid) (*CrossMsgs, bool, error) {
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply ipc_types/ipc_subnet_actor.go:78]: done

File: ipc_types/types.go (hunk 2 above)
Comment (ranchalp) [reply ipc_types/types.go:17]: ok

File: ipc_types/types.go (hunk 3 above)
Comment (ranchalp) [reply ipc_types/types.go:75]: Done

File: ipc_types/types.go (hunk 4 above)
Comment (ranchalp) [reply ipc_types/types.go:86]: Done.

File: ipc_types/types.go (hunk 5 above)
Comment (ranchalp) [reply ipc_types/types.go:97]: Outdated to do.

File: ipc_types/ipc_subnet_actor.go (hunk 1 above)
Comment (ranchalp) [reply ipc_types/ipc_subnet_actor.go:170]: I did not implement those rust methods as they were only public within the crate. What is an example of (even if updated) implementation of a cid() method for a struct like one of these?

File: types/ipc_subnet_actor.go [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+	Consensus         ConsensusType
+	MinValidatorStake abi.TokenAmount
+	MinValidators     uint64
+	FinalityThreshold ChainEpoch
+	CheckPeriod       ChainEpoch
+	Genesis           []byte
+}
+
+
+pub fn cid(&self) -> Cid {
```
Comment (adlrocha) [inline types/ipc_subnet_actor.go:122]: remove?

File: types/types.go [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+	"strings"
+)
+
+type IPCAddress struct {
+	SubnetID   SubnetID
+	RawAddress address.Address
+}
+
+func (i IPCAddress) String() string {
+	// `-` is used as delimiter instead of `/`
```
Comment (adlrocha) [inline types/types.go:19]: remove

File: types/types.go [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+	Actor  address.Address
+}
+
+var id0, _ = address.NewIDAddress(0) //TODO check with Alfonso
+
+const (
+	RootStr          = "/root"
+	SubnetSeparator  = "/"
+	UndefStr         = ""
+	IPCAddrSeparator = "-"
```
Comment (adlrocha) [inline types/types.go:52]: the separator is `:` not `-`. https://github.com/consensus-shipyard/ipc-actors/blob/877177fa3f56ead824da53084b9957619e823d59/sdk/src/address.rs#L9

File: types/types.go [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+
+// RootSubnet is the ID of the root network
+var RootSubnet = SubnetID{
+	Parent: RootStr,
+	Actor:  id0,
+}
+
+// UndefSubnetID is the undef ID
+var UndefSubnetID = SubnetID{
+	Parent: UndefStr,
```
Comment (adlrocha) [inline types/types.go:63]: This is not `UndefStr` but `SubnetSeparator`: https://github.com/consensus-shipyard/ipc-actors/blob/877177fa3f56ead824da53084b9957619e823d59/sdk/src/subnet_id.rs#L24

**Review by adlrocha — APPROVED** [review]
Hey, @ranchalp, do you mind addressing the latest comments and getting this merged? Also, if you could get CI configured? (we can use github actiosn directly). We will keep iterating with the code in follow-up PRs if needed. Thanks!

**Pull request merged**